Initial draft of battery post (#349)
memfault/interrupt: _drafts/monitoring-battery-life.md

---
title: Understanding Battery Performance of IoT Devices
description: Monitoring & predicting the battery life of millions of embedded devices in production is made possible through the use of state of charge delta metrics and aggregations.

> author: tyler

Author: Oh! Forgot about them - they aren't in the `_template.md` so I forgot about them.

Will add some!

Having all three of these power consumption profiles help paint a picture of how much power this piece of hardware may consume, and therefore how quickly it might drain a battery.

### Power Profiles for each component

The first step is to create baselines for how much power each component will consume **over time**. Although the spec sheets from the hardware vendor are great and will generally tell you power consumption - they aren’t always accurate, and different components from different batches will have different power usage characteristics.

The general steps to accomplishing this are:

- Ensure that your development boards have power rails so that you can isolate many of the components and you can easily attach probes to them.
- Get a nice multimeter that can measure µA and mA.
- Write a special firmware that instruments a single component in various power modes, and determine the current draw for each mode. This should be the simplest possible firmware and ideally only contain driver code.
- Rinse and repeat with every component that could have a big impact on power consumption.

> For example, the accelerometer was a component in a wearable device that could consume a lot of power if left in the wrong state for periods of time. We wrote a firmware that would set it to different sampling rates and record the current consumption and then used this to determine at what sampling rate we could keep the accelerometer while still achieving our desired 7-day battery life.

Reviewer: ```suggestion
For example, the accelerometer is a component in a wearable device that could consume a lot of power if left in the wrong state for periods of time. We wrote firmware that would set it to different sampling rates and record the current consumption and then used this to determine at what sampling rate we could keep the accelerometer while still achieving our desired 7-day battery life.
```

Also the last sentence here runs on a bit. Consider restructuring.

Author: Thanks. Will update.

### State of Charge (SoC) over Time

Once we determine whether or not our hardware can successfully meet our minimum requirements for battery life, now we need to put this to the test from the other direction. The primary thing we’ll look into now is how to measure the state of charge (SoC) of the battery over time to ensure that the device can last as long as required.

One of the most common and simplest ways to measure the current capacity within a battery, or the SoC of a device, is by measuring the voltage of the battery during a known and consistent power draw, but it's not the only way.

#### Coulomb Counting with Fuel Gauges

> At Pebble, we had a fuel gauge on one of the watches. A fuel gauge is a nifty hardware component that can indicate the battery’s SoC and health. It can understand the battery’s current capacity in two ways: measuring the voltage from the battery and learning the patterns over time, and also Coulomb counting, which measures how much current passes in and out of the battery over time.

Reviewer: Not sure if I'm reading this wrong, but isn't that three ways it measures batter capacity?

Author: I'll make it cleaner

If you happen to be lucky enough to be able to use a fuel gauge and happen to be using Zephyr RTOS, definitely check out their new Fuel Gauge API[^zephyr_fuel], announced recently in Zephyr v3.3.

#### State of charge with voltage

This is most commonly what companies have, as it’s very simple to get from the battery system and doesn’t consume any extra power. The problem with only tracking the voltage is that it’s not easily human-understandable, does not increase or drop linearly, and will change under operating conditions, as mentioned above.

<p align="center">
  <img width="400" src="{% img_url monitoring-battery-life/battery-curve.png %}" alt="battery-curve" />
</p>

We need something better if possible.

#### State of charge with voltage and battery curve

One thing that can be done to help convert voltage to a percentage is to come up with a battery curve. A battery curve is simple: it's a map between a battery's voltage and the relative percentage that likely pertains to that voltage. Products also usually have both a charge and a discharge curve.

> A battery curve is what companies will have after they have some form of understanding of their battery properties and have spent some time. It is more easily understood by customer support teams and engineers that aren’t directly involved with the battery subsystem. That isn’t to say that it

Reviewer: `have spent some time`, not sure what should follow this. `have spent some time using the battery`?

Author: Oof, fixed.

A nice tool that I came across this year at Embedded World was Qoitech[^qoitech], which builds a product to help users build charge and discharge curves[^qoitech_curves] under different environments. I believe their product is well worth the money if it can help companies translate a cryptic voltage reading to a percentage that everyone can understand.

## Brief Primer on Metrics

Before continuing to the following sections about capturing and aggregating metrics around battery life, let’s briefly discuss what a metric is, since I continue to come across firmware engineers who haven’t thought about them all that much.

> A **metric** is a measurement captured at runtime, and combining large numbers of metrics and calculating statistics is called an **aggregation**.

Author: ```suggestion
A **metric** is a measurement captured at runtime, and the process of combining large numbers of metrics and calculating statistics is known as **aggregation**.
```

But we aren't here to talk about Bluetooth disconnections, let's talk about power!

## Metrics that typically contribute to power consumption

Here's a semi-exhaustive list of items that I've tracked in the past that would help paint a picture for me and my colleagues about what was consuming battery life.

### Connectivitiy & networking

This would cover any sort of wireless radio, such as Wi-Fi, LoRa, Bluetooth, ZigBee, LTE, etc.

- packets sent & received
- bytes sent & received
- number of connections and disconnections
- time spent in each state (advertising, connecting)
- radio strength & power settings
- throughput
- number of retry attempts

### Peripheral usage

Here, we try to measure any peripheral that might consume significant amounts of power.

- sensor on/off time (acceleromter, gyroscope, GPS, cameras, compass, etc.)
- actuator on/off time
- actuator total distance
- display & backlight on/off time
- number of display refreshes
- camera on/off time
- storage read/writes/erases

### CPU & Code usage

- CPU awake, sleep, and deep-sleep time
- time spent running each task in the system
- time running power-hungry blocks of code
- boot time

> - number of logs written

Author: Ooh! Never thought about this one! Added.
